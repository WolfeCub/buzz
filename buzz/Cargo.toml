[package]
name = "buzz"
version = "0.1.0"
edition = "2021"

[dependencies]
buzz-types = { path = "../buzz-types" }
buzz-codegen = { path = "../buzz-codegen" }

[dev-dependencies]
proptest = "1"
