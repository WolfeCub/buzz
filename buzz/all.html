<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../buzz/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../buzz/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location">Crate buzz</h2></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../buzz/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><div class="theme-picker hidden"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="22" height="22" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></form></nav></div><section id="main-content" class="content"><h1 class="fqn"><span class="in-band">List of all items</span></h1><h3 id="Structs">Structs</h3><ul class="structs docblock"><li><a href="json/struct.Json.html">json::Json</a></li><li><a href="prelude/struct.Buzz.html">prelude::Buzz</a></li><li><a href="prelude/struct.BuzzContext.html">prelude::BuzzContext</a></li><li><a href="prelude/struct.Inject.html">prelude::Inject</a></li><li><a href="prelude/struct.InjectMut.html">prelude::InjectMut</a></li></ul><h3 id="Enums">Enums</h3><ul class="enums docblock"><li><a href="json/enum.JsonValue.html">json::JsonValue</a></li></ul><h3 id="Traits">Traits</h3><ul class="traits docblock"><li><a href="prelude/trait.Respond.html">prelude::Respond</a></li></ul><h3 id="Macros">Macros</h3><ul class="macros docblock"><li><a href="prelude/macro.routes.html">prelude::routes</a></li></ul><h3 id="Attribute-Macros">Attribute Macros</h3><ul class="attributes docblock"><li><a href="prelude/attr.delete.html">prelude::delete</a></li><li><a href="prelude/attr.get.html">prelude::get</a></li><li><a href="prelude/attr.options.html">prelude::options</a></li><li><a href="prelude/attr.patch.html">prelude::patch</a></li><li><a href="prelude/attr.post.html">prelude::post</a></li><li><a href="prelude/attr.put.html">prelude::put</a></li></ul><h3 id="Derive-Macros">Derive Macros</h3><ul class="derives docblock"><li><a href="prelude/derive.Deserialize.html">prelude::Deserialize</a></li></ul></section><section id="search" class="content hidden"></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="buzz" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.61.0 (fe5b13d68 2022-05-18)" ></div>
</body></html>