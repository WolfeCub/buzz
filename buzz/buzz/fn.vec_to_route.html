<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../buzz/prelude/fn.vec_to_route.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../buzz/prelude/fn.vec_to_route.html">../../buzz/prelude/fn.vec_to_route.html</a>...</p>
    <script>location.replace("../../buzz/prelude/fn.vec_to_route.html" + location.search + location.hash);</script>
</body>
</html>