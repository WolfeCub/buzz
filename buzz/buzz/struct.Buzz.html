<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../buzz/prelude/struct.Buzz.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../buzz/prelude/struct.Buzz.html">../../buzz/prelude/struct.Buzz.html</a>...</p>
    <script>location.replace("../../buzz/prelude/struct.Buzz.html" + location.search + location.hash);</script>
</body>
</html>