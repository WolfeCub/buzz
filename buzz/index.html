<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A rust web framework that avoids dependancies wherever possible."><meta name="keywords" content="rust, rustlang, rust-lang, buzz"><title>buzz - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../buzz/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../buzz/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Crate buzz</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.1.0</li><li><a id="all-types" href="all.html">All Items</a></li></div></ul><section><div class="block"><ul><li><a href="#modules">Modules</a></li></ul></div></section><div id="sidebar-vars" data-name="buzz" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../buzz/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><div class="theme-picker hidden"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="22" height="22" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">buzz</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/buzz/lib.rs.html#1-138">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A rust web framework that avoids dependancies wherever possible.</p>
<h2 id="examples"><a href="#examples">Examples</a></h2>
<p>These examples are best viewed as cargo docs <a href="https://wolfecub.github.io/buzz/">here</a>.</p>
<p>Here we return a <a href="https://doc.rust-lang.org/1.61.0/std/primitive.str.html" title="&amp;str"><code>&amp;str</code></a> and it behaves as expected. We can register a route by using
the <a href="prelude/struct.Buzz.html#method.routes"><code>routes</code></a> method and the <a href="prelude/macro.routes.html"><code>routes!</code></a> macro.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/foo&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">foo</span>() -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="string">&quot;foo&quot;</span>
}

<span class="kw">fn</span> <span class="ident">main</span>() {
    <span class="ident">Buzz::new</span>(<span class="string">&quot;127.0.0.1:8080&quot;</span>)
        .<span class="ident">routes</span>(<span class="macro">routes!</span>(<span class="ident">foo</span>))
        .<span class="ident">run_server</span>();
}</code></pre></div>
<p><a href="https://doc.rust-lang.org/1.61.0/core/option/enum.Option.html" title="Option"><code>Option</code></a> variants work as long as the inner element implements <a href="prelude/trait.Respond.html"><code>Respond</code></a>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/it&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">it</span>() -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="prelude-val">Some</span>(<span class="string">&quot;it&quot;</span>)
}</code></pre></div>
<p>Returning <a href="https://doc.rust-lang.org/1.61.0/core/option/enum.Option.html#variant.None"><code>None</code></a> gives you a 404.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/empty&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">empty</span>() -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="prelude-ty">Option</span>::<span class="op">&lt;</span>()<span class="op">&gt;</span><span class="ident">::None</span>
}</code></pre></div>
<p>Returning an the <code>Err</code> variant of <a href="https://doc.rust-lang.org/1.61.0/core/result/enum.Result.html" title="Result"><code>Result</code></a> gives you a 500 with the content <code>Err</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/other&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">other</span>() -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="prelude-ty">Result</span>::<span class="op">&lt;</span>(), <span class="kw">_</span><span class="op">&gt;</span><span class="ident">::Err</span>(<span class="ident">std::fmt::Error::default</span>())
}</code></pre></div>
<p>Query parameters are just regular arguments. We can specify anything that implements <a href="https://doc.rust-lang.org/1.61.0/core/str/traits/trait.FromStr.html" title="std::str::FromStr"><code>std::str::FromStr</code></a></p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/params/{hello}/{number}&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">params</span>(<span class="ident">hello</span>: <span class="ident">String</span>, <span class="ident">number</span>: <span class="ident">i32</span>) -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="macro">format!</span>(<span class="string">&quot;{hello} {number}&quot;</span>)
}</code></pre></div>
<p>Query params are also just args but are <a href="https://doc.rust-lang.org/1.61.0/core/option/enum.Option.html" title="Option"><code>Option</code></a>.
In the case of <code>?arg=something</code> arg here will be <code>Some(&quot;something&quot;)</code>
and <a href="https://doc.rust-lang.org/1.61.0/core/option/enum.Option.html#variant.None"><code>None</code></a> if <code>arg</code> is not present in the query string.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/query&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">query</span>(<span class="ident">arg</span>: <span class="prelude-ty">Option</span><span class="op">&lt;</span><span class="ident">String</span><span class="op">&gt;</span>) -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="ident">arg</span>
}</code></pre></div>
<p><a href="prelude/struct.BuzzContext.html"><code>BuzzContext</code></a> can be injected anywhere and get extra info about the request.
Here we grab an arbitrary header to make use of.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/context&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">context</span>(<span class="ident">context</span>: <span class="ident">BuzzContext</span>) -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="ident">context</span>.<span class="ident">headers</span>.<span class="ident">get</span>(<span class="string">&quot;Some-Header&quot;</span>).<span class="ident">map</span>(<span class="op">|</span><span class="ident">h</span><span class="op">|</span> (<span class="kw-2">*</span><span class="ident">h</span>).<span class="ident">to_owned</span>())
}</code></pre></div>
<p>Registered params can be injected anywhere with the special type <a href="prelude/struct.Inject.html"><code>Inject</code></a>.
You may also inject it mutably using <a href="prelude/struct.InjectMut.html"><code>InjectMut</code></a>. Note that this will
block other callers from acquiring another mutable reference until this one is released.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/inject&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">inject</span>(<span class="ident">val</span>: <span class="ident">Inject</span><span class="op">&lt;</span><span class="ident">i32</span><span class="op">&gt;</span>) -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="ident">val</span>.<span class="ident">to_string</span>()
}

<span class="attribute">#[<span class="ident">get</span>(<span class="string">&quot;/inject-mut&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">inject_mut</span>(<span class="kw-2">mut</span> <span class="ident">val</span>: <span class="ident">InjectMut</span><span class="op">&lt;</span><span class="ident">i32</span><span class="op">&gt;</span>) -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="kw-2">*</span><span class="ident">val</span> <span class="op">=</span> <span class="number">77</span>;
}

<span class="kw">fn</span> <span class="ident">main</span>() {
    <span class="ident">Buzz::new</span>(<span class="string">&quot;127.0.0.1:8080&quot;</span>)
        .<span class="ident">routes</span>(<span class="macro">routes!</span>(<span class="ident">inject</span>, <span class="ident">inject_mut</span>))
        <span class="comment">// Here we register a type that can be injected</span>
        .<span class="ident">register</span>::<span class="op">&lt;</span><span class="ident">i32</span><span class="op">&gt;</span>(<span class="number">42</span>)
        .<span class="ident">run_server</span>();
}</code></pre></div>
<p>Here we derive <a href="../buzz_types/traits/deserialize/trait.Deserialize.html"><code>Deserialize</code></a> for <code>Task</code> allowing us to map json or other
data types to it. We tell buzz which request param to inject the body for by specifying <code>body = &quot;arg_name&quot;</code>.
Buzz knows to deserialize the incoming request as JSON since we wrapped our variable in the <a href="json/struct.Json.html"><code>Json</code></a> type.
<a href="https://doc.rust-lang.org/1.61.0/core/option/enum.Option.html" title="Option"><code>Option</code></a> fields can either be absent or <code>null</code>. Both will deserialize to <a href="https://doc.rust-lang.org/1.61.0/core/option/enum.Option.html#variant.None"><code>None</code></a></p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Deserialize</span>)]</span>
<span class="kw">struct</span> <span class="ident">Task</span> {
    <span class="ident">index</span>: <span class="ident">i64</span>,
    <span class="ident">content</span>: <span class="ident">String</span>,
    <span class="ident">extra</span>: <span class="prelude-ty">Option</span><span class="op">&lt;</span><span class="ident">String</span><span class="op">&gt;</span>,
}

<span class="attribute">#[<span class="ident">post</span>(<span class="string">&quot;/json&quot;</span>, <span class="ident">body</span> <span class="op">=</span> <span class="string">&quot;request_body&quot;</span>)]</span>
<span class="kw">fn</span> <span class="ident">json</span>(<span class="ident">request_body</span>: <span class="ident">Json</span><span class="op">&lt;</span><span class="ident">Task</span><span class="op">&gt;</span>) -&gt; <span class="kw">impl</span> <span class="ident">Respond</span> {
    <span class="macro">format!</span>(<span class="string">&quot;{}. {}&quot;</span>, <span class="ident">request_body</span>.<span class="ident">index</span>, <span class="ident">request_body</span>.<span class="ident">content</span>)
}</code></pre></div>
</div></details><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left import-item"><code>pub use <a class="mod" href="../buzz_types/index.html" title="mod buzz_types">buzz_types</a> as types;</code></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left import-item"><code>pub use buzz_codegen as codegen;</code></div><div class="item-right docblock-short"></div></div></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="json/index.html" title="buzz::json mod">json</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="prelude/index.html" title="buzz::prelude mod">prelude</a></div><div class="item-right docblock-short"></div></div></div></section><section id="search" class="content hidden"></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="buzz" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.61.0 (fe5b13d68 2022-05-18)" ></div>
</body></html>