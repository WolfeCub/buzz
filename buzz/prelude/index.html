<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `prelude` mod in crate `buzz`."><title>buzz::prelude - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-46132b98.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="buzz" data-themes="" data-resource-suffix="" data-rustdoc-version="1.85.1 (4eb161250 2025-03-15)" data-channel="1.85.1" data-search-js="search-75f5ac3e.js" data-settings-js="settings-0f613d39.js" ><script src="../../static.files/storage-59e33391.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-5f194d8c.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../buzz/index.html">buzz</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module prelude</a></h2><h3><a href="#macros">Module Items</a></h3><ul class="block"><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li><li><a href="#derives" title="Derive Macros">Derive Macros</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate buzz</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><span class="rustdoc-breadcrumbs"><a href="../index.html">buzz</a></span><h1>Module <span>prelude</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/buzz/lib.rs.html#142">Source</a> </span></div><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.routes.html" title="macro buzz::prelude::routes">routes</a></div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Buzz.html" title="struct buzz::prelude::Buzz">Buzz</a></div></li><li><div class="item-name"><a class="struct" href="struct.BuzzContext.html" title="struct buzz::prelude::BuzzContext">Buzz<wbr>Context</a></div><div class="desc docblock-short">Holds metadata about the incoming <code>HttpRequest</code> that’s being handled.</div></li><li><div class="item-name"><a class="struct" href="struct.Inject.html" title="struct buzz::prelude::Inject">Inject</a></div><div class="desc docblock-short">Wrapper that indicates a shared reference should be injected.</div></li><li><div class="item-name"><a class="struct" href="struct.InjectMut.html" title="struct buzz::prelude::InjectMut">Inject<wbr>Mut</a></div><div class="desc docblock-short">Wrapper that indicates an exclusive reference should be injected.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Respond.html" title="trait buzz::prelude::Respond">Respond</a></div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.delete.html" title="attr buzz::prelude::delete">delete</a></div></li><li><div class="item-name"><a class="attr" href="attr.get.html" title="attr buzz::prelude::get">get</a></div></li><li><div class="item-name"><a class="attr" href="attr.options.html" title="attr buzz::prelude::options">options</a></div></li><li><div class="item-name"><a class="attr" href="attr.patch.html" title="attr buzz::prelude::patch">patch</a></div></li><li><div class="item-name"><a class="attr" href="attr.post.html" title="attr buzz::prelude::post">post</a></div></li><li><div class="item-name"><a class="attr" href="attr.put.html" title="attr buzz::prelude::put">put</a></div></li></ul><h2 id="derives" class="section-header">Derive Macros<a href="#derives" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="derive" href="derive.Deserialize.html" title="derive buzz::prelude::Deserialize">Deserialize</a></div></li></ul></section></div></main></body></html>