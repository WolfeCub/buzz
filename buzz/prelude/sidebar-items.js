initSidebarItems({"attr":[["delete",""],["get",""],["options",""],["patch",""],["post",""],["put",""]],"derive":[["Deserialize",""]],"macro":[["routes",""]],"struct":[["Buzz",""],["BuzzContext","Holds metadata about the incoming `HttpRequest` thatâ€™s being handled."],["Inject",""],["InjectMut",""]],"trait":[["Respond",""]]});