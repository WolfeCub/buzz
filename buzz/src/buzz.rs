use std::collections::HashMap;
use std::error::Error;
use std::io::prelude::*;
use std::net::TcpListener;
use std::net::TcpStream;

use crate::http_parse::*;
use buzz_types::*;

/* TODO: Use enum in the handler map rather than strings */
pub struct Buzz {
    addr: &'static str,
    handlers: HashMap<RouteMetadata<'static>, fn() -> HttpResponse>,
}

impl Buzz {
    pub fn new(addr: &'static str) -> Self {
        Self {
            addr,
            handlers: HashMap::new(),
        }
    }

    pub fn route(mut self, route: (fn() -> HttpResponse, RouteMetadata<'static>)) -> Self {
        self.handlers.insert(route.1, route.0);
        self
    }

    pub fn run_server(&self) {
        let listener = TcpListener::bind(self.addr).unwrap();

        for stream in listener.incoming() {
            let stream = stream.unwrap();

            match self.handle_connection(stream) {
                Ok(_) => {}
                Err(e) => panic!("{}", e),
            }
        }
    }

    fn handle_connection(&self, mut stream: TcpStream) -> Result<(), Box<dyn Error>> {
        let mut buffer = [0; 1024];

        stream.read(&mut buffer)?;

        let request = parse_http(&buffer)?;

        let metadata = RouteMetadata {
            method: &request.method.to_string(),
            path: &request.path,
        };
        let response = match self.handlers.get(&metadata) {
            Some(handler) => handler(),
            None => HttpResponse::new(HttpStatusCode::NotFound),
        };

        write_response(&mut stream, &response)?;

        stream.flush()?;
        stream.shutdown(std::net::Shutdown::Both)?;

        Ok(())
    }
}

fn write_response(stream: &mut TcpStream, request: &HttpResponse) -> std::io::Result<()> {
    /* TODO: Not hardcoded version. What do we actually support? */
    stream.write(
        format!(
            "HTTP/1.1 {} {}\r\n",
            to_status_num(request.status_code),
            request.status_code.to_string()
        )
        .as_bytes(),
    )?;

    for (key, value) in &request.headers {
        stream.write(format!("{}: {}\r\n", key, value).as_bytes())?;
    }

    stream.write(b"\r\n")?;

    /* TODO: Buffer? */
    if let Some(body) = &request.body {
        stream.write(body.as_bytes())?;
    }
    stream.flush()?;

    Ok(())
}
