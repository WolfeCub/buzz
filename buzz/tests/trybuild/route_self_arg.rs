use buzz::prelude::*;

#[get("/self-arg")]
fn self_arg(self) {
}

fn main() {}
