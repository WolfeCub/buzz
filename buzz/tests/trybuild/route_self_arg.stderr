error: macros that expand to items must be delimited with braces or followed by a semicolon
 --> tests/trybuild/route_self_arg.rs:3:1
  |
3 | #[get("/self-arg")]
  | ^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `get` (in Nightly builds, run with -Z macro-backtrace for more info)
help: change the delimiters to curly braces
  |
3 | {}
  | ~
help: add a semicolon
  |
3 | #[get("/self-arg")];
  |                    +

error: Found self in args which is not allowed
 --> tests/trybuild/route_self_arg.rs:3:1
  |
3 | #[get("/self-arg")]
  | ^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `get` (in Nightly builds, run with -Z macro-backtrace for more info)
